Add Konnected component with support for discovery, binary sensor, and switch

heythisisnate opened the pull request:
## Description:
Adds a component for [Konnected](https://konnected.io) devices. Konnected is an [open source](https://github.com/konnected-io/konnected-security) application that runs on a NodeMCU ESP8266. Konnected develops hardware designed for connecting traditional wired alarm system sensors and sirens to home automation.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation:** https://github.com/home-assistant/home-assistant.github.io/pull/5102

## Example entry for `configuration.yaml`:
```yaml
konnected:
  access_token: REPLACE_ME_WITH_A_RANDOM_STRING
  devices:
    - id: 8bcd53
      binary_sensors:
        - zone: 1
          type: door
          name: 'Front Door'
        - zone: 3
          type: motion
          name: 'Test Motion'
      switches:
        - zone: out
          name: siren
    - id: 438a38
      binary_sensors:
        - pin: 1
          type: motion
          name: 'Office Motion'
        - pin: 2
          type: door
          name: 'Office Door'
      switches:
        - pin: 5
          name: 'Garage Door'
          activation: low
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.


[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

homeassistant commented:
Hi @heythisisnate, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#13670).

Once you do that we will be able to review and accept this pull request.

Thanks!

heythisisnate committed:
Add Konnected component with support for discovery, binary sensor, and switch

Co-authored-by: Eitan Mosenkis <eitan@mosenkis.net>

heythisisnate commented:
This is my first Home Assistant PR. This code was written in collaboration with @emosenkis (squashed and cleaned up commits for easy reading). Can anyone help us out with a review?

fabaff commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,293 @@
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: {
+            'auth_token': str,
```
Looks like that this is required. We have a validation helper for strings in `homeassistant.helpers.config_validation`.

fabaff commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,293 @@
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: {
```
The `DOMAIN` contains a schema as well.

fabaff commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,293 @@
+        DOMAIN: {
+            'auth_token': str,
+            vol.Optional('home_assistant_url'): str,
+            'devices': [{
```
Please check `const.py` if there is already a defined constant.

fabaff commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,293 @@
+    if cfg is None:
+        cfg = {}
+
+    auth_token = cfg.get('auth_token') or 'supersecret'
```
Is 'supersecret' the default? If yes, then pass it in the validation.

heythisisnate replied:
I opted to remove the default and make it required for the user to set.

heythisisnate committed:
Use more built-in constants from const.py

heythisisnate commented:
Thank you for the feedback @fabaff. I think I've addressed all of the concerns mentioned.

pfiggins commented:
Anyone know the typical timeline for approval and merging of a PR? I would like to use this integration but am on Hass.io at the moment.

heythisisnate commented:
I just rebased on the latest `dev` branch and fixed a merge conflict. Is there anything else that I need to here @fabaff?

heythisisnate committed:
Fix switch actuation with low-level trigger

pfiggins commented:
It would be great to get this accepted and merged in ASAP. Konnected is the whole reason I'm beginning to migrate from SmartThings!

dejanzelic commented:
I'm excited for this module! I tried it out and it works flawlessly!

OttoWinter commented on homeassistant/components/binary_sensor/konnected.py:
```diff
@@ -0,0 +1,75 @@
+        self._device_id = device_id
+        self._pin_num = pin_num
+        self._state = self._data.get(ATTR_STATE)
+        self._device_class = self._data.get(CONF_TYPE, 'motion')
```
This already seems to be set by the configuration schema:

```python
vol.Required(CONF_TYPE, default='motion'): DEVICE_CLASSES_SCHEMA,
```

OttoWinter commented on homeassistant/components/binary_sensor/konnected.py:
```diff
@@ -0,0 +1,75 @@
+        self._name = self._data.get(CONF_NAME, 'Konnected {} Zone {}'.format(
+            device_id, PIN_TO_ZONE[pin_num]))
+        self._data['entity'] = self
+        _LOGGER.info('Created new sensor: %s', self._name)
```
This seems overly verbose for an `info` log message... I would consider it a debug message if at all.

OttoWinter commented on homeassistant/components/binary_sensor/konnected.py:
```diff
@@ -0,0 +1,75 @@
+    sensors = [KonnectedBinarySensor(device_id, pin_num, pin_data)
+               for pin_num, pin_data in
+               data[CONF_DEVICES][device_id][CONF_SENSORS].items()]
+    async_add_devices(sensors, True)
```
The binary sensor doesn't seem to have an `update` method, so passing `True` here has no effect and should be removed.

OttoWinter commented on homeassistant/components/binary_sensor/konnected.py:
```diff
@@ -0,0 +1,75 @@
+        self._state = state
+        self._data[ATTR_STATE] = state
+        self.async_schedule_update_ha_state()
+        _LOGGER.info('Updating state: %s is %s', self.name, state)
```
This debug message also seems quite over-verbose. The `async_schedule_update_ha_state()` will already make a state changed debug message for you.

OttoWinter commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,298 @@
+                vol.Required(CONF_ID, default=''): config_validation.string,
+                vol.Optional(CONF_SENSORS): [{
+                    vol.Exclusive(CONF_PIN, 's_pin'): vol.Any(*PIN_TO_ZONE),
+                    vol.Exclusive(CONF_ZONE, 's_pin'): vol.Any(*ZONE_TO_PIN),
```
If the user doesn't provide `pin` nor `zone`, this config validation will still pass even though it's clearly invalid. See [`has_at_least_one_key`](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/config_validation.py#L49) for potentially fixing this.

heythisisnate replied:
Nice. thanks for that tip.

OttoWinter commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,298 @@
+                vol.Optional(CONF_SENSORS): [{
+                    vol.Exclusive(CONF_PIN, 's_pin'): vol.Any(*PIN_TO_ZONE),
+                    vol.Exclusive(CONF_ZONE, 's_pin'): vol.Any(*ZONE_TO_PIN),
+                    vol.Required(CONF_TYPE, default='motion'):
```
If it's required, why does it have a default?

OttoWinter commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,298 @@
+                    vol.Exclusive(CONF_PIN, 'a_pin'): vol.Any(*PIN_TO_ZONE),
+                    vol.Exclusive(CONF_ZONE, 'a_pin'): vol.Any(*ZONE_TO_PIN),
+                    vol.Optional(CONF_NAME): config_validation.string,
+                    vol.Required('activation', default='high'):
```
Same as above, required keys should not have a default.

OttoWinter commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,298 @@
+        port = info.get(CONF_PORT)
+
+        device = KonnectedDevice(hass, host, port, cfg)
+        device.setup()
```
This is a sync call within the event loop. From what I see `setup()` calls `sync_device()` which calls `client.put_settings`. I presume the latter is a sync HTTP call. But when inside the event loop (`asyncio.coroutine`, `async_`) you can only do async calls or otherwise you will block the entire Home Assistant code execution.

heythisisnate replied:
What's the best way to resolve this? In this case we don't care about waiting for a response from `device.setup()`. Is there a simple way to "fire and forget"?

OttoWinter replied:
```python
hass.async_add_job(device.setup)
```

OttoWinter commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,298 @@
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {'auth_token': auth_token}
+
+    @asyncio.coroutine
```
Please use the new `async`/`await` syntax if you can. We have Python 3.5 now and I've seen a few PRs getting comments like this one. See http://stackabuse.com/python-async-await-tutorial/ for a simple tutorial.

heythisisnate replied:
Ok, I will read up on it. I'm pretty new to Python so I'm sure that I just modeled this after some other code that maybe wasn't up date.

OttoWinter commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,298 @@
+        }
+
+        if 'devices' not in self.hass.data[DOMAIN]:
+            self.hass.data[DOMAIN][CONF_DEVICES] = {}
```
Just a style comment: it's weird to read code where sometimes `'devices'` is used and sometimes `CONF_DEVICES`. I would recommend using one of them and sticking to it IMO, but it's not required.

OttoWinter commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,298 @@
+        if 'devices' not in self.hass.data[DOMAIN]:
+            self.hass.data[DOMAIN][CONF_DEVICES] = {}
+
+        _LOGGER.info('Storing data in hass.data[konnected]: %s', device_data)
```
This also seems like a non-info log message. Especially with `info` messages, I don't expect to see raw python dictionaries in the logs.

OttoWinter commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,298 @@
+        _LOGGER.info('%s: desired actuator config: %s', self.device_id,
+                     desired_actuator_config)
+        _LOGGER.info('%s: current actuator config: %s', self.device_id,
+                     current_actuator_config)
```
These log messages should IMO not be `info`. Home Assistant's default is to print every log message above and including `info`. If someone needs these log messages they can still manually set the log level for this component using the [logger component](https://www.home-assistant.io/components/logger/).

heythisisnate replied:
You're right, these were really meant for debugging discovery. I will change them to `debug`.

OttoWinter commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,298 @@
+        data = hass.data[DOMAIN]
+
+        auth = request.headers.get(AUTHORIZATION, None)
+        if 'Bearer {}'.format(self.auth_token) != auth:
```
This is not safe to timing attacks. See [`hmac.compare_digest`](https://github.com/home-assistant/home-assistant/blob/321eb2ec6f077191e20a7fc07490a18df6f11f3f/homeassistant/components/http/auth.py#L34) for safe auth token checks.

OttoWinter commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,298 @@
+    extra=vol.ALLOW_EXTRA,
+)
+
+DEPENDENCIES = ['http']
```
If this integration requires `discovery:` to be enabled as the docs state, it might be good to have it in `DEPENDENCIES` too I think.

heythisisnate committed:
Quiet logging; Improve schema validation.

heythisisnate committed:
Execute sync request outside of event loop

heythisisnate committed:
Whitespace cleanup

heythisisnate commented:
@OttoWinter thank you for the feedback, very helpful. I think I've addressed everything. Could you take another look?

OttoWinter commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,304 @@
+            vol.Required('auth_token'): config_validation.string,
+            vol.Required(CONF_DEVICES): [{
+                vol.Required(CONF_ID, default=''): config_validation.string,
+                vol.Optional(CONF_SENSORS): [_SENSOR_SCHEMA],
```
I'm not sure about this configuration format, when I see `sensors` in a Home Assistant configuration I expect the key to refer to actual (analog) [sensors](https://www.home-assistant.io/components/sensor/), not [binary sensors](https://www.home-assistant.io/components/binary_sensor/). Maybe change the key to `CONF_BINARY_SENSORS`?

heythisisnate replied:
Yes, makes sense to use consistent terminology.

OttoWinter commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,304 @@
+        DOMAIN: vol.Schema({
+            vol.Required('auth_token'): config_validation.string,
+            vol.Required(CONF_DEVICES): [{
+                vol.Required(CONF_ID, default=''): config_validation.string,
```
I think this should be an `vol.Optional`. I find it quite weird that `vol.Required` also has a `default` option. `vol.Optional` makes more sense semantically I think if there is a default and AFAIK Home Assistant has generally used default values with `Optional`.

heythisisnate replied:
ID should be required. I'll remove the default.

OttoWinter commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,304 @@
+            vol.Required(CONF_DEVICES): [{
+                vol.Required(CONF_ID, default=''): config_validation.string,
+                vol.Optional(CONF_SENSORS): [_SENSOR_SCHEMA],
+                vol.Optional(CONF_SWITCHES): [_SWITCH_SCHEMA],
```
Home Assistant tries to be quite lenient with its config in quite a few places, that includes automatically converting single-element dictionaries to lists. For example, a configuration like this should preferably be allowed:

```yaml
konnected:
  # ...
  devices:
    - id: something
      switches:
        pin: 1
        name: Switch1
```

This can be achieved with:

```python
vol.Optional(CONF_SWITCHES): vol.All(config_validation.ensure_list, [_SWITCH_SCHEMA]),
```

also, `homeassistant.helpers.config_validation` is usually imported as `cv` for a shorthand:

```python
from homeassistant.helpers import config_validation as cv
```

OttoWinter commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,304 @@
+        if user_config:
+            _LOGGER.debug('Configuring Konnected device %s', self.device_id)
+            self.save_data()
+            self.hass.async_add_job(self.sync_device)
```
Now this is potentially an async call from a sync context. I know it's only called from inside the event loop right now, but every method that can be directly called from an async context must be prefixed `async_` and be either a coroutine or `@callback`. See https://developers.home-assistant.io/docs/en/asyncio_categorizing_functions.html

heythisisnate replied:
This means that the `setup` function and `sync_device` should be annotated as async as well, right? I've categorized them both as callbacks because they don't have to wait for any result. Please let me know if I'm thinking about this wrong.

heythisisnate committed:
Cleanup config validation; async device setup

heythisisnate commented:
@OttoWinter thanks again for the thorough review. I've addressed everything in the latest commit and also rebased from the current upstream `dev` branch. Please let me know if there's anything else.

For users who are using this branch -- please change the `sensors` key in your configuration.yml to `binary_sensors`. I'll update the documentation.

jschwalbe commented:
Very excited for this to be official!

BarisEris commented:
Folks as an active user of Konnected, I'm eager to see this PR fully processed -- is there any outstanding task that is holding this back? Thanks.

riteshp83 commented:
Is there a time frame for when Konnected will be included directly in HomeAssistant?

sateeshkumarj commented:
Eagerly waiting for Konnected to see official..

ryanchrisw commented:
Excited to see this PR get merged! It's been running flawless since it was first announced that Konnected had Home Assistant support.

v0ltage120 commented:
Please homeassistant approve this PR!!! Konnected device is a revolution in home automation regarding alarm system recycling!

heythisisnate committed:
update API endpoint for Konnected 2.2.0 changes

OttoWinter commented:
@heythisisnate Yeah the asyncio stuff can be pretty difficult to learn (see [this](https://developers.home-assistant.io/docs/en/asyncio_index.html) for a basic intro).

To speed up getting this into Home Assistant, I now copied your code and modified it in a gist [here](https://gist.github.com/OttoWinter/9522ad0419b50c1a5e57d8280b190e30). I haven't tried it locally because I don't use konnected, so I would recommend you trying it first.

Otherwise, this integration looks pretty good to me 👍

heythisisnate committed:
update async coroutines via @OttoWinter

heythisisnate commented:
@OttoWinter thank you so much! Yes, that was very helpful. I think the syntax was tripping me up. I've done async programming before with Javascript promises, but this new syntax reads like it's synchronous so it was a little confusing. I'm starting to grok it now.

Anyway, I tested your changes from the gist and it works. I just committed those changes. I hope this can be the last change for this initial release, as I know a lot of people are waiting on this. Thanks again, this has been educational.

What's the process like for merging and getting this on to the release schedule?

heythisisnate committed:
make backwards compatible with Konnected < 2.2.0

heythisisnate commented:
hey @OttoWinter @fabaff,

I don't mean to be pushy as I know very well how open source projects go, but another week has gone by and I feel like I'm still in the dark about what the next steps are to get this component merged into the next release. Is there anything left for me to do here? Please advise.

I'm getting emails every day from Konnected users looking for an ETA. I'm not sure what to tell them 😕 

We've just finished shipping all of our [Kickstarter orders](https://www.kickstarter.com/projects/nateclark/konnected-alarm-panel-revive-your-wired-alarm-syst) for the Konnected Alarm Panel, so now there are literally thousands of Konnected devices out there in the world just waiting for HASS to talk to them! 

By the way, I just pushed a small compatibility fix and rebased on to the current `dev` branch. This should be good to go 🤞

kylegordon commented:
Hey @balloob @syssi @pvizeli 
Is this something you could cast your eyes over and consider for merging into dev?
Cheers!

syssi commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,312 @@
+        vol.Exclusive(CONF_PIN, 'a_pin'): vol.Any(*PIN_TO_ZONE),
+        vol.Exclusive(CONF_ZONE, 'a_pin'): vol.Any(*ZONE_TO_PIN),
+        vol.Optional(CONF_NAME): cv.string,
+        vol.Optional('activation', default='high'):
```
Please use a const. `CONF_ACTIVATION`

syssi commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,312 @@
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema({
+            vol.Required('auth_token'): cv.string,
```
Same here.

syssi commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,312 @@
+    if cfg is None:
+        cfg = {}
+
+    auth_token = cfg.get('auth_token')
```
Use the new const.

syssi commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,312 @@
+        self.user_config = config
+
+        import konnected
+        self.client = konnected.Client(host, str(port))
```
What happens if the device is unavailable?

heythisisnate replied:
The device depends on the `discovery` component to be recognized and initialized when HASS starts up, so if the device is unavailable at startup, it would never reach this code.

It is possible that the device is online at HASS startup but then later goes offline -- in this case it obviously would no longer push state updates to HASS, and signals from HASS to trigger switches would timeout. I think this is not a big problem for now. In the future I'm thinking of adding some retry logic and possibly device health/status with notification if a device goes offline. I think that's too much (potentially unnecessary) complexity for this first release though.

syssi commented on homeassistant/components/switch/konnected.py:
```diff
@@ -0,0 +1,84 @@
+
+    def turn_on(self, **kwargs):
+        """Send a command to turn on the switch."""
+        self._client.put_device(self._pin_num, int(self._activation == 'high'))
```
'high' and 'low' should be a const, too.

syssi commented:
I would be happy if you care about introducing some consts as last step. It's a very nice integration and should be approved afterwards!

bradleyscott commented:
Assuming the constants, Would that mean it could be released as part of 0.70 @syssi?

syssi commented:
I hope so. :-P I believe the merge window closes on friday.

pfiggins commented:
Awesome! Nate has been great about turning the fixes around quickly. So it sounds like we'll be getting native integration in 0.70. Thank you @syssi!

heythisisnate committed:
add constants suggested by @syssi

heythisisnate commented:
@syssi Thank you SO much for the review! 🙌 
I've added the constants as suggested. Please have a second look.

*Konnected users who are beta testing this:* I've changed the `auth_token` configuration variable to `access_token` to be consistent with other similar HASS configs and to reuse an existing constant string. I'm updating the documentation accordingly.

syssi commented:
@heythisisnate Sorry, but you've to move the const to the top of your component. The const.py is just used for prominent consts. ;-)

heythisisnate commented:
@syssi Oh, gotcha. I misunderstood. Give me 5 minutes.

heythisisnate commented:
@syssi updated

syssi commented:
Do you like to add your component and name to the CODEOWNERS file? It basically means you'll automatically get notified for all changes in your source files.

heythisisnate committed:
add to CODEOWNERS

heythisisnate commented:
@syssi Nice. Added.

syssi reviewed (approved):
LGTM. There are some more consts which should be introduced on the long run IMO.

Kane610 commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,318 @@
+    def save_data(self):
+        """Save the device configuration to `hass.data`.
+
+        TODO: This can probably be refactored and tidied up.
```
Since this is a new component we should make sure to have it as ready as possible. At the least remove todo comment, at best follow the todo 👍

heythisisnate committed:
remove TODO comment

heythisisnate commented:
Thank you for the approval @syssi. So what happens next? Do I have to merge this or does somebody from the core team do the merge into `dev`?

MartinHjelmare commented on homeassistant/components/binary_sensor/konnected.py:
```diff
@@ -0,0 +1,74 @@
+import logging
+
+from homeassistant.components.binary_sensor import BinarySensorDevice
+from homeassistant.components.konnected import (DOMAIN, PIN_TO_ZONE)
```
We're not allowed to overwrite the binary sensor domain with the `konnected` component domain. Instead import it like so:
```py
from homeassistant.components.konnected import (DOMAIN as KONNECTED_DOMAIN, ...)
```

MartinHjelmare commented on homeassistant/components/binary_sensor/konnected.py:
```diff
@@ -0,0 +1,74 @@
+        """Return the device class."""
+        return self._device_class
+
+    @asyncio.coroutine
```
We're now using Python 3.5 async syntax, ie `async def`.

MartinHjelmare commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,315 @@
+
+        import konnected
+        self.client = konnected.Client(host, str(port))
+        self.status = self.client.get_status()
```
This call is making a request, ie I/O, and we're inside a coroutine here. That's not allowed. Probably change `async_device_discovered` to be non async as well as `KonnectedDevice`. If the interface library is not async, there's usually no point in making the home assistant component async.

heythisisnate replied:
Hmm, I think this was all originally synchronous but at the suggestion of another reviewer I made these into coroutines: https://github.com/home-assistant/home-assistant/pull/13670#discussion_r183218555
So now I'm thoroughly confused as to the best way to resolve. Seems like either I make everything synchronous or change the `konnected` library to use asyncio http. Which is preferred? Pros/cons of each?

MartinHjelmare commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,315 @@
+        if (desired_sensor_configuration != current_sensor_configuration) or \
+                (current_actuator_config != desired_actuator_config):
+            _LOGGER.debug('pushing settings to device %s', self.device_id)
+            self.client.put_settings(
```
This is also making a request inside a coroutine.

MartinHjelmare commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,315 @@
+    def sensor_configuration(self):
+        """Return the configuration map for syncing sensors."""
+        return [{'pin': p} for p in
+                self.stored_configuration[CONF_BINARY_SENSORS].keys()]
```
`.keys()` is not needed when iterating a dict.

MartinHjelmare commented on homeassistant/components/konnected.py:
```diff
@@ -0,0 +1,315 @@
+            return self.json_message('uninitialized sensor/actuator',
+                                     status_code=HTTP_INTERNAL_SERVER_ERROR)
+
+        await entity.async_set_state(state)
```
Entities should usually not be accessed directly from outside their platform. That's against the design philosophy of home assistant. Instead, we have the dispatch helper for this purpose, to send an update from the component to platform entities.

It's ok to store a key that references a specific device or entity, eg the entity_id in a dict to be able to send the update to the correct target with the dispatcher. But we shouldn't store the entities themselves so that they are accessible from outside the platform.

heythisisnate replied:
How do I get the entity_id to store in a dict? It's unclear to me where this is set or how to retrieve it. `self.entity_id` returns `None` from inside the entity class. I must be missing something here.

MartinHjelmare replied:
You might not need the entity_id. You don't need it specifically to use the dispatcher. I just mentioned it as an example identifier of an entity.

There's two parts to the dispatcher, a connect function and a send function. First you *connect* a target function to a target key. Then you *send* an update to the target key which will call all connected target functions of the target key.

The target key can be anything that is specific enough for what you want to target. If you want to target a specific entity, the entity_id is probably good. You could also use the builtin `id` function on the entity instance to get a unique id of the entity instance, if you don't need the key to be persistent between restarts.

You want to connect the entity target method from the entity coroutine method `async_added_to_hass`. At this point, the entity_id will be available on the entity, ie it will be something other than `None`.

heythisisnate replied:
I understand conceptually.
Could you point me to a component that uses this pattern? This would help me tremendously.

MartinHjelmare commented on homeassistant/components/switch/konnected.py:
```diff
@@ -0,0 +1,87 @@
+    def _set_state(self, state):
+        self._state = state
+        self._data[ATTR_STATE] = state
+        self.schedule_update_ha_state()
```
Is there no feedback from the device when it changes state? If there is no feedback, the entity should be marked to use `assumed_state`, ie set that entity property to `True`.

If there is feedback after a state change, we shouldn't update state from the `turn_on`/`turn_off` methods, but let the feedback handle that, if it's done within a couple of seconds.

MartinHjelmare reviewed (commented):
Please open a new PR, to at least address the I/O being done inside coroutines.
